### PR TITLE
token-2022: add calculate post and pre fee amounts

### PR DESCRIPTION
**Pull request by samkim-crypto**
An attempt at simplifying transfer fee calculation.

- Simplified fee calculation using ceiled division `ceil[ a / b ]` as `(a + b - 1) / b`.
- Added `calculate_post_fee_amount` and `calculate_pre_fee_amount`.

It seems like for most applications, computing (1) the net transfer amount from the gross amount or (2) the gross amount from the net amount is very useful.

@joncinque What do you think? The updated `calculate_inverse_fee` breaks a couple of the edge case tests, but it seems to maintain the precision. It also seems like `calculate_post_fee_amount` and `calculate_pre_fee_amount` can be used directly for the swap program.

### CODE REVIEW

File: token/program-2022/src/extension/transfer_fee/mod.rs [hunk 1]
```diff
@@ -32,63 +35,71 @@ pub struct TransferFee {
     pub transfer_fee_basis_points: PodU16,
 }
 impl TransferFee {
+    /// Calculate ceiled division
```
Comment (joncinque) [inline token/program-2022/src/extension/transfer_fee/mod.rs:38]: nit: could you provide a little more info in the comment about this implementation?

File: token/program-2022/src/extension/transfer_fee/mod.rs [hunk 2]
```diff
@@ -32,63 +35,71 @@ pub struct TransferFee {
...
+                .try_into() // guaranteed to be okay
+                .ok()?;
+
             Some(cmp::min(raw_fee, u64::from(self.maximum_fee)))
         }
     }
 
-    /// Calculate the fee that would produce the given output
-    pub fn calculate_inverse_fee(&self, post_fee_amount: u64) -> Option<u64> {
+    /// Calculate the net transfer amount after deducting fees
```
Comment (joncinque) [inline token/program-2022/src/extension/transfer_fee/mod.rs:61]: nit, right?
```suggestion
    /// Calculate the gross transfer amount after deducting fees
```

File: token/program-2022/src/extension/transfer_fee/mod.rs [hunk 3]
```diff
@@ -300,10 +311,10 @@ pub(crate) mod test {
...
         let maximum_fee = u64::from(transfer_fee.maximum_fee);
         // hit maximum fee
-        assert_eq!(
-            maximum_fee,
-            transfer_fee.calculate_inverse_fee(u64::MAX).unwrap()
-        );
+        // assert_eq!(
+        //     maximum_fee,
+        //     transfer_fee.calculate_inverse_fee(u64::MAX).unwrap()
+        // );
```
Comment (joncinque) [inline token/program-2022/src/extension/transfer_fee/mod.rs:317]: Can you use a different value that works to hit the maximum fee here?

File: token/program-2022/src/extension/transfer_fee/mod.rs [hunk 4]
```diff
@@ -32,63 +35,71 @@ pub struct TransferFee {
...
+    ///
+    /// The original transfer amount may not always exist on large net transfer amounts due to
+    /// overflow. In this case, `None` is returned.
+    /// e.g. The net fee amount is `u64::MAX` with a positive fee rate.
+    pub fn calculate_pre_fee_amount(&self, post_fee_amount: u64) -> Option<u64> {
+        let maximum_fee = u64::from(self.maximum_fee);
         let transfer_fee_basis_points = u16::from(self.transfer_fee_basis_points) as u128;
         if transfer_fee_basis_points == 0 {
+            Some(post_fee_amount)
+        } else if transfer_fee_basis_points == ONE_IN_BASIS_POINTS || post_fee_amount == 0 {
```
Comment (joncinque) [inline token/program-2022/src/extension/transfer_fee/mod.rs:82]: If post fee amount is 0, that doesn't necessarily mean that the pre fee amount is 0. If I transfer 1 token and the fee is 1bp, that'll have a fee of 1.

File: token/program-2022/src/extension/transfer_fee/mod.rs [hunk 5]
```diff
@@ -351,7 +362,7 @@ pub(crate) mod test {
             transfer_fee.calculate_inverse_fee(one).unwrap()
         );
         // zero is 1 token
-        assert_eq!(minimum_fee, transfer_fee.calculate_inverse_fee(0).unwrap());
+        // assert_eq!(minimum_fee, transfer_fee.calculate_inverse_fee(0).unwrap());
```
Comment (joncinque) [inline token/program-2022/src/extension/transfer_fee/mod.rs:365]: This should still work too

**Review by joncinque — COMMENTED** [review]
Great job! Mainly some nits.

I'm glad that the extra ceiling bit is totally unnecessary with your improved calc.

File: token/program-2022/src/extension/transfer_fee/mod.rs (hunk 4 above)
Comment (samkim-crypto) [reply token/program-2022/src/extension/transfer_fee/mod.rs:82]: Yeah, so this is something that I thought about for a bit and is related to one of the commented out tests below as well. The gross amount is not uniquely determined due to rounding. So I thought the function should naturally provide the minimum gross amount that will produce a specified net amount.

Suppose a sender wants a receiver to receive a certain number of tokens after fees. Then it could use this function to calculate how much to send in total before fees are deducted. Then it is natural to always choose the smaller amount since the sender does not want to overpay when possible.

If the post fee amount is 0, then both 0 and 1 should work. If the fee rate is 10,000, then any amount below max fee should also produce 0. In all these cases, the amount 0 is the minimum pre fee amount that will produce 0 post fee amount.

File: token/program-2022/src/extension/transfer_fee/mod.rs (hunk 5 above)
Comment (samkim-crypto) [reply token/program-2022/src/extension/transfer_fee/mod.rs:365]: Yep, this is related to the comment above. Assume that bp is positive. If post fee amount is 0, then `calculate_pre_fee_amount` could produce either 0 or 1, which results in fee of either 0 or 1.

Currently, `calculate_pre_fee_amount` produces pre fee of 0 on post fee 0, so this test should fail since `minimum_fee` is 1. If the current behavior of `calculate_pre_fee_amount` is fine, then `minimum_fee` should be replaced by 0 in this test.

File: token/program-2022/src/extension/transfer_fee/mod.rs (hunk 4 above)
Comment (joncinque) [reply token/program-2022/src/extension/transfer_fee/mod.rs:82]: We had a discussion about this offline, and you're right, we'll go with this implementation!

File: token/program-2022/src/extension/transfer_fee/mod.rs (hunk 5 above)
Comment (joncinque) [reply token/program-2022/src/extension/transfer_fee/mod.rs:365]: Per our discussion, let's have this be 0 then!

File: token/program-2022/src/extension/transfer_fee/mod.rs [hunk 6]
```diff
@@ -302,7 +316,9 @@ pub(crate) mod test {
         // hit maximum fee
         assert_eq!(
             maximum_fee,
-            transfer_fee.calculate_inverse_fee(u64::MAX).unwrap()
+            transfer_fee
+                .calculate_inverse_fee(u64::MAX - maximum_fee)
```
Comment (joncinque) [inline token/program-2022/src/extension/transfer_fee/mod.rs:320]: perfect choice!

**Review by joncinque — APPROVED** [review]
Thanks for improving this
